In [2]:
!pip install apyori

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5976 sha256=34fbc0b359db9ab1a0121d12f3a4698789698488430fd8d0bac3639706a397bf
  Stored in directory: /root/.cache/pip/wheels/32/2a/54/10c595515f385f3726642b10c60bf788029e8f3a1323e3913a
Successfully built apyori


In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from apyori import apriori

In [4]:
#Loading the dataset
store_data = pd.read_csv('store_data.csv',header=None)

In [5]:
store_data.head()

0          1           2                 3             4   \
0         shrimp    almonds     avocado    vegetables mix  green grapes   
1        burgers  meatballs        eggs               NaN           NaN   
2        chutney        NaN         NaN               NaN           NaN   
3         turkey    avocado         NaN               NaN           NaN   
4  mineral water       milk  energy bar  whole wheat rice     green tea   

                 5     6               7             8             9   \
0  whole weat flour  yams  cottage cheese  energy drink  tomato juice   
1               NaN   NaN             NaN           NaN           NaN   
2               NaN   NaN             NaN           NaN           NaN   
3               NaN   NaN             NaN           NaN           NaN   
4               NaN   NaN             NaN           NaN           NaN   

               10         11     12     13             14      15  \
0  low fat yogurt  green tea  honey  salad  mineral water  salmon   
1             NaN        NaN    NaN    NaN            NaN     NaN   
2             NaN        NaN    NaN    NaN            NaN     NaN   
3             NaN        NaN    NaN    NaN            NaN     NaN   
4             NaN        NaN    NaN    NaN            NaN     NaN   

                  16               17       18         19  
0  antioxydant juice  frozen smoothie  spinach  olive oil  
1                NaN              NaN      NaN        NaN  
2                NaN              NaN      NaN        NaN  
3                NaN              NaN      NaN        NaN  
4                NaN              NaN      NaN        NaN

In [6]:
store_data.shape

(7501, 20)

In [7]:
# Converting the dataset into list of lists since apriori works with list of lists only

records = []

for i in range(0,7501):
    records.append([str(store_data.values[i,j]) for j in range(0,20)])

In [8]:
records[1]

['burgers',
 'meatballs',
 'eggs',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan',
 'nan']

transaction in which the bread was purchased / total transactions

In [11]:
# Building the first apriori model
association_rules = apriori(records, min_support=0.0045, min_confidence=0.2, min_lift=3, min_length=2)
association_results = list(association_rules)

In [12]:
print(len(association_results))

48


In [13]:
# Glance at the first rule
print(association_results[0])

RelationRecord(items=frozenset({'chicken', 'light cream'}), support=0.004532728969470737, ordered_statistics=[OrderedStatistic(items_base=frozenset({'light cream'}), items_add=frozenset({'chicken'}), confidence=0.29059829059829057, lift=4.84395061728395)])


It shows that 'chicken' and 'light cream' are frequently bought together in the dataset. The rule applies to only 0.45% of the transactions and has a confidence of 0.29, meaning that when a customer buys 'light cream', there is a 29% chance that they will also buy 'chicken'. The lift of the rule is 4.84, indicating that customers who buy both 'chicken' and 'light cream' are 4.84 times more likely to buy them together than customers who buy only one of these items.

In [14]:
print(association_results[1])

RelationRecord(items=frozenset({'escalope', 'mushroom cream sauce'}), support=0.005732568990801226, ordered_statistics=[OrderedStatistic(items_base=frozenset({'mushroom cream sauce'}), items_add=frozenset({'escalope'}), confidence=0.3006993006993007, lift=3.790832696715049)])


It shows that 'escalope' and 'mushroom cream sauce' are frequently bought together in the dataset. The rule applies to only 0.57% of the transactions and has a confidence of 0.30, meaning that when a customer buys 'mushroom cream sauce', there is a 30% chance that they will also buy 'escalope'. The lift of the rule is 3.79, indicating that customers who buy both 'escalope' and 'mushroom cream sauce' are 3.79 times more likely to buy them together than customers who buy only one of these items.

In [16]:
# Building the second apriori model
association_rules = apriori(records, min_support=0.010, min_confidence=0.5, min_lift=2)
association_results = list(association_rules)

In [18]:
print(len(association_results))

4


In [19]:
print(association_results[0])

RelationRecord(items=frozenset({'eggs', 'mineral water', 'ground beef'}), support=0.010131982402346354, ordered_statistics=[OrderedStatistic(items_base=frozenset({'eggs', 'ground beef'}), items_add=frozenset({'mineral water'}), confidence=0.5066666666666667, lift=2.125563012677107)])


It shows that 'eggs', 'mineral water', and 'ground beef' are frequently bought together in the dataset. The rule applies to only 1.01% of the transactions and has a confidence of 0.51, meaning that when a customer buys both 'eggs' and 'ground beef', there is a 51% chance that they will also buy 'mineral water'. The lift of the rule is 2.13, indicating that customers who buy all three items together are 2.13 times more likely to buy them together than customers who buy only two of these items.

In [20]:
print(association_results[1])

RelationRecord(items=frozenset({'milk', 'mineral water', 'ground beef'}), support=0.011065191307825623, ordered_statistics=[OrderedStatistic(items_base=frozenset({'milk', 'ground beef'}), items_add=frozenset({'mineral water'}), confidence=0.503030303030303, lift=2.1103077757440176)])


It shows that 'milk', 'mineral water', and 'ground beef' are frequently bought together in the dataset. The rule applies to only 1.11% of the transactions and has a confidence of 0.50, meaning that when a customer buys both 'milk' and 'ground beef', there is a 50% chance that they will also buy 'mineral water'. The lift of the rule is 2.11, indicating that customers who buy all three items together are 2.11 times more likely to buy them together than customers who buy only two of these items.